lib/tree.rb

```diff
@@ -16,51 +16,114 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log n)
+  # Space Complexity: O(1)
   def add(key, value)
-    raise NotImplementedError
+    @root = add_helper(@root, key, value)
+  end
+
+  def add_helper(root, key, value)
+    if root.nil?
+      root = TreeNode.new(key, value)
+      return root
+    end
+
+    if root.key > key
+      root.left = add_helper(root.left, key, value)
+    else
+      root.right = add_helper(root.right, key, value)
+    end
+
+    root
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log n)
+  # Space Complexity: O(log n)
   def find(key)
-    raise NotImplementedError
+    find_helper(@root, key)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def find_helper(node, key)
+    return nil if node.nil?
+
+    if node.key == key
+      node.value
+    elsif node.key < key
+      find_helper(node.right, key)
+    else
+      find_helper(node.left, key)
+    end
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def inorder
-    raise NotImplementedError
+    result = []
+    inorder_helper(@root, result)
+    result
+  end
+
+  def inorder_helper(node, result)
+    return if node.nil?
+
+    inorder_helper(node.left, result)
+    result.push({ key: node.key, value: node.value })
+    inorder_helper(node.right, result)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def preorder
-    raise NotImplementedError
+    result = []
+    preorder_helper(@root, result)
+    result
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def preorder_helper(node, result)
+    return if node.nil?
+
+    result.push({ key: node.key, value: node.value })
+    preorder_helper(node.left, result)
+    preorder_helper(node.right, result)
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def postorder
-    raise NotImplementedError
+    result = []
+    postorder_helper(@root, result)
+    result
+  end
+
+  def postorder_helper(node, result)
+    return if node.nil?
+
+    postorder_helper(node.left, result)
+    postorder_helper(node.right, result)
+    result.push({ key: node.key, value: node.value })
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def height
```

Comment:
👍 The space complexity (due to recursion) is O(log n) if the tree is balanced and O(n) if it's unbalanced.